src/Datadog.Trace/Util/PropertyFetcher.cs

```diff
@@ -0,0 +1,98 @@
+        /// <param name="propertyName">The name of the property that this instance will fetch.</param>
+        public PropertyFetcher(string propertyName)
+        {
+            _propertyName = propertyName;
+        }
+
+        /// <summary>
+        /// Gets the value of the property on the specified object.
+        /// </summary>
+        /// <param name="obj">The object that contains the property.</param>
+        /// <returns>The value of the property on the specified object.</returns>
+        public object Fetch(object obj)
+        {
+            Type objType = obj.GetType();
+
+            if (objType != _expectedType)
+            {
+                TypeInfo typeInfo = objType.GetTypeInfo();
+                var propertyInfo = typeInfo.DeclaredProperties.FirstOrDefault(p => string.Equals(p.Name, _propertyName, StringComparison.InvariantCultureIgnoreCase));
+                _fetchForExpectedType = PropertyFetch.FetcherForProperty(propertyInfo);
+                _expectedType = objType;
+            }
+
+            return _fetchForExpectedType.Fetch(obj);
+        }
+
+        /// <summary>
+        /// PropertyFetch is a helper class. It takes a PropertyInfo and then knows how
+        /// to efficiently fetch that property from a .NET object (See Fetch method).
+        /// It hides some slightly complex generic code.
+        /// </summary>
+        private class PropertyFetch
+        {
+            /// <summary>
+            /// Create a property fetcher from a .NET Reflection <see cref="PropertyInfo"/> class that
+            /// represents a property of a particular type.
+            /// </summary>
+            /// <param name="propertyInfo">The property that this instance will fetch.</param>
+            /// <returns>The new property fetcher.</returns>
+            public static PropertyFetch FetcherForProperty(PropertyInfo propertyInfo)
+            {
+                if (propertyInfo == null)
+                {
+                    // returns null on any fetch.
+                    return new PropertyFetch();
+                }
+
+                Type typedPropertyFetcher = typeof(TypedFetchProperty<,>);
+                Type instantiatedTypedPropertyFetcher = typedPropertyFetcher.GetTypeInfo().MakeGenericType(
+                    propertyInfo.DeclaringType, propertyInfo.PropertyType);
+
+                return (PropertyFetch)Activator.CreateInstance(instantiatedTypedPropertyFetcher, propertyInfo);
+            }
+
+            /// <summary>
+            /// Gets the value of the property on the specified object.
+            /// </summary>
+            /// <param name="obj">The object that contains the property.</param>
+            /// <returns>The value of the property on the specified object.</returns>
+            public virtual object Fetch(object obj)
+            {
+                return null;
+            }
+
+            private class TypedFetchProperty<TObject, TProperty> : PropertyFetch
```

Comment:
Looking at these changes in the DiagnosticsSource, if we need to get properties off objects that are value types, we'll have to re-write the `PropertyFetcher` ever so slightly. Can you check that all the objects we pass to PropertyFetcher (whose properties we try to get) are all reference types? https://github.com/dotnet/runtime/commit/13b9d5412c6f0ee4f1882cfa4986d9d5652a95e9

Reply:
Good find. All the current properties are reference types.